mlflow/utils/rest_utils.py

```diff
@@ -14,27 +12,91 @@
 from mlflow.utils.string_utils import strip_suffix
 from mlflow.exceptions import MlflowException, RestException
 
-_REST_API_PATH_PREFIX = "/api/2.0"
 RESOURCE_DOES_NOT_EXIST = "RESOURCE_DOES_NOT_EXIST"
-
-_logger = logging.getLogger(__name__)
-
+_REST_API_PATH_PREFIX = "/api/2.0"
 _DEFAULT_HEADERS = {"User-Agent": "mlflow-python-client/%s" % __version__}
+# Response codes that generally indicate transient network failures and merit client retries,
+# based on guidance from cloud service providers
+# (https://docs.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#general-rest-and-retry-guidelines)
+_TRANSIENT_FAILURE_RESPONSE_CODES = frozenset(
+    [
+        408,  # Request Timeout
+        429,  # Too Many Requests
+        500,  # Internal Server Error
+        502,  # Bad Gateway
+        503,  # Service Unavailable
+        504,  # Gateway Timeout
+    ]
+)
+
+
+def _get_http_response_with_retries(
+    method, url, max_retries, backoff_factor, retry_codes, **kwargs
+):
+    """
+    Performs an HTTP request using Python's `requests` module with an automatic retry policy.
+
+    :param method: a string indicating the method to use, e.g. "GET", "POST", "PUT".
+    :param url: the target URL address for the HTTP request.
+    :param max_retries: Maximum total number of retries.
+    :param backoff_factor: a time factor for exponential backoff. e.g. value 5 means the HTTP
+      request will be retried with interval 5, 10, 20... seconds. A value of 0 turns off the
+      exponential backoff.
+    :param retry_codes: a list of HTTP response error codes that qualifies for retry.
+    :param kwargs: Additional keyword arguments to pass to `requests.Session.request()`
+
+    :return: requests.Response object.
+    """
+    assert 0 <= max_retries < 10
+    assert 0 <= backoff_factor < 120
+    retry = Retry(
+        total=max_retries,
+        connect=max_retries,
+        read=max_retries,
+        redirect=max_retries,
+        status=max_retries,
+        status_forcelist=retry_codes,
+        allowed_methods=None,  # Allow retry on any HTTP methods.
```

Comment:
allowed_methods doesn't work on wheel install either. I have to change it back to method_whitelist:

MlflowException: API request to https://test-shard-jin-zhang.dev.databricks.com/api/2.0/mlflow/runs/create failed with exception __init__() got an unexpected keyword argument 'allowed_methods'

Reply:
I see, I think we're using an old version of `urrlib3` in databricks. Can we use `allowed_methods` if the version of `urllib3` is >= 1.26.0?

```python
from packaging.version import Version

methods_arg = "allowed_methods" if Version(urllib3.__version__) >= Version("1.26.0") else "method_whitelist"
Retry(..., **{methods_arg: None})
```

Reply:
@harupy I installed urllib3 1.26.6, and I don't see `allowed_methods` defined on the `Retry()` API. I think it's a documentation issue.

Reply:
```
In [4]: import urllib3

In [5]: urllib3.__version__
Out[5]: '1.26.6'

In [6]: from urllib3.util import Retry

In [7]: help(Retry)

class Retry(builtins.object)
 |  Retry(total=10, connect=None, read=None, redirect=None, status=None, other=None, allowed_methods=<object object at 0x7fd1f09eec30>, status_forcelist=None, backoff_factor=0, raise_on_redirect=True, raise_on_status=True, history=None, respect_retry_after_header=True, remove_headers_on_redirect=<object object at 0x7fd1f09eec30>, method_whitelist=<object object at 0x7fd1f09eec30>)
```

Reply:
@dbczumar Are you really sure that `Retry` doesn't have `allowed_methods`?

```python
>>> import urllib3
>>> urllib3.__version__
'1.26.6'
>>> from urllib3.util import Retry
>>> Retry()
Retry(total=10, connect=None, read=None, redirect=None, status=None)
>>> Retry(allowed_methods=None)
Retry(total=10, connect=None, read=None, redirect=None, status=None)

>>> Retry(method_whitelist=None)
__main__:1: DeprecationWarning: Using 'method_whitelist' with Retry is deprecated and will be removed in v2.0. Use 'allowed_methods' instead
Retry(total=10, connect=None, read=None, redirect=None, status=None)
```

Reply:
https://github.com/urllib3/urllib3/blob/1.26.6/src/urllib3/util/retry.py#L233

Reply:
Ah, missed it. Looks like it's there. I'll add this conditional logic.